### PR TITLE
fix(explore): Prevent shared controls from checking feature flags outside React render

### PR DESCRIPTION
**Pull request by codyml**
### SUMMARY

Feature flags are saved to a global window object and read using the `isFeatureEnabled` utility function, which returns `true` or `false` after checking the global window object.  As noted in #21201, this process is prone to error because feature flags are set at JS module execution time and sometimes also read at JS module execution time, creating a race condition that can be triggered by a change in webpack bundling.  This PR takes steps towards resolving one instance of this issue by moving the feature flag check to component render.  Previously, shared chart controls were exported as an object of control config objects, each of which contained the control's React component to render, params used by the control-rendering engine, and props that the control's React component would be rendered with.  For each shared control that had drag-and-drop functionality, a different config object would be exported at JS module execution time if `isFeatureEnabled` checks indicated that drag-and-drop functionality was enabled.  This PR moves the `isFeatureEnabled` check to inside a wrapper component, and uses a single config to render that wrapper component regardless of feature flag status.  While there was some overlap between config objects, the DnD and non-DnD versions of some control components required different props, so some refactoring was required to make it possible to merge the two config objects' props to form one unified config object.

### TESTING INSTRUCTIONS

- Confirm that every chart type's control panel works as expected when the `ENABLE_EXPLORE_DRAG_AND_DROP` feature flag is enabled and disabled.  When `ENABLE_EXPLORE_DRAG_AND_DROP` is enabled, also ensure that everything works when `ENABLE_DND_WITH_CLICK_UX` is enabled and disabled.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Required feature flags: `ENABLE_EXPLORE_DRAG_AND_DROP`, `ENABLE_DND_WITH_CLICK_UX`
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by kgabryje** [comment]
Wow that's a tremendous piece of work @codyml!
@jinghua-qa Could you please help testing this? The PR basically affects all controls in Explore - we need to make sure that drag and dropping works as it did before

File: superset-frontend/src/explore/components/controls/DndColumnSelectControl/DndMetricSelect.tsx
Comment (kgabryje) [inline superset-frontend/src/explore/components/controls/DndColumnSelectControl/DndMetricSelect.tsx:129]: We definitely should improve typing here... (not in this PR, that's a general note)

File: superset-frontend/packages/superset-ui-chart-controls/src/shared-controls/dndControls.tsx
Comment (kgabryje) [inline superset-frontend/packages/superset-ui-chart-controls/src/shared-controls/dndControls.tsx:50]: Could you add a comment that will explain the idea behind using props from non-dnd component in dnd control definition? Might be useful for other contributors in the future

**Comment by kgabryje** [comment]
Looks fantastic! Though I can't wait to revert all this when we fully get rid of non-dnd interface 😄
I've tested viz types with non-standard controls (e.g. Table chart, deckgl charts) in dnd and non-dnd mode and it all seems to work well.
I'm going to hold off with approving though until we get a green light from QA

**Comment by jinghua-qa** [comment]
I found an issue that for world map, after save for changes, the chart viz will be empty, i am not able to reproduce in master branch, it may be this PR releated.

https://user-images.githubusercontent.com/81597121/189720113-9f21cb2a-cf92-4949-9372-e7876dec9952.mov

**Comment by andrey-zayats** [comment]
> I found an issue that for world map, after save for changes, the chart viz will be empty, i am not able to reproduce in master branch, it may be this PR releated.
> 
>  Screen.Recording.2022-09-12.at.10.22.05.AM.mov

I can reproduce this issue with different types of charts, but only with a certain dataset "wb_health_population"

**Comment by andrey-zayats** [comment]
QA verified. PR LGTM

**Comment by jinghua-qa** [comment]
> I found an issue that for world map, after save for changes, the chart viz will be empty, i am not able to reproduce in master branch, it may be this PR releated.
> 
>  Screen.Recording.2022-09-12.at.10.22.05.AM.mov
This issue is able to reproduce in other superset ephemeral, but not in superset master. It is not related to this PR.  LGTM!

**Review by jinghua-qa — APPROVED** [review]

File: superset-frontend/packages/superset-ui-chart-controls/src/shared-controls/dndControls.tsx
Comment (zhaoyongjie) [inline superset-frontend/packages/superset-ui-chart-controls/src/shared-controls/dndControls.tsx:101]: @codyml the `options` in the controls was an Object data structure that key as column name and value was metadata of column, this means the `options[column_name]` would return the metadata of column. This part modified the `options` to the Array instead of the Object, so the caller can't use `options[column_name]` to get a column.

Comment (codyml) [reply superset-frontend/packages/superset-ui-chart-controls/src/shared-controls/dndControls.tsx:101]: Yes, I did this because the DnD and non-DnD versions of some shared control components had different interfaces: the DnD versions expected an object while the non-DnD versions expected an array.  So, to merge the props the two versions needed to expect the same thing, and the simplest solution seemed to be to change that so they both expected an array because IIRC it was used as an array more than it was used as an object in both.  Let me know if you think another solution would have worked better and I can make that change.

Comment (zhaoyongjie) [reply superset-frontend/packages/superset-ui-chart-controls/src/shared-controls/dndControls.tsx:101]: thanks for the explanation. both work for me.

File: superset-frontend/packages/superset-ui-chart-controls/src/types.ts
Comment (justinpark) [inline superset-frontend/packages/superset-ui-chart-controls/src/types.ts:240]: @codyml why do you make this parameter optional?
This breaks all our custom override type checking.

Comment (codyml) [reply superset-frontend/packages/superset-ui-chart-controls/src/types.ts:240]: Hmm, I think that was a quick fix to stop the histogram viz plugin from complaining after I updated it to use shared controls.  I opened [a PR](https://github.com/apache/superset/pull/22014) that reverts that change and instead updates the plugin – does that look like a better solution?  Would be great to spin up an ephemeral env to check if there are any side-effects for Histogram charts.

Comment (codyml) [reply superset-frontend/packages/superset-ui-chart-controls/src/types.ts:240]: (Assuming it passes CI)

Comment (codyml) [reply superset-frontend/packages/superset-ui-chart-controls/src/types.ts:240]: https://github.com/apache/superset/pull/22014 was merged, let me know if that solves the issue!

**Pull request merged**
